### ISSUE TITLE
Create a browsable gallary of all images by Scriberia

### ISSUE THREAD

**Issue by malvikasharan** [the-turing-way/the-turing-way#2046]
### Summary

* We have all the Scriberia illustrations on our GitHub folder, but browsing them is not that easy, and therefore many people may not know what illustrations exist that they can use or refer to: https://github.com/alan-turing-institute/the-turing-way/tree/master/book/website/figures
* I wonder if someone knows how to visualise all the images from a folder in GitHub as an album. Any template or example will be really helpful.

### What needs to be done?

- [x] Select a method - see/add comments
  * @nehamoopen is working on an implementation using RShiny 
- [ ] For image visualisation, select the best way to look at all the images:
  - Separate gallery view <-- current plan!
  - A dedicated location at the end of the book
  - Something else?

More tasks, copied from https://github.com/the-turing-way/the-turing-way/issues/2046#issuecomment-2456751902 and other comments

- [ ] Keyword search
- [ ] Alt-text 
    * Not all images have alt text and it probably makes sense to create the gallery without going back and fixing every one of those. Maybe there's some filler alt text that says "This image does not have alt text - we are working on improving our accessibility." ?
- [ ] Download button 
- [ ] Scriberia attribution 
- [ ] License 
- [ ] Auto resize images to fit screen
- [ ] Click through buttons for images (back and forth, like slides)
- [ ] provide images with and without text for download
- [ ] show a random illustration
- [ ] consider where it will need to be hosted, whether the app needs to run and deployed on servers
   * webassembly allows compiling scripts in the browser

#### Roadmap

Scroll down for the rough Roadmap! See: https://github.com/the-turing-way/the-turing-way/issues/2046#issuecomment-2465299057

#### Open questions

- [ ] Does it need to be built in Python instead of R? Or perhaps we use a totally different tool/platform like some of the suggestions in the previous discussion?
- [ ] Will it be embedded in the book or hosted elsewhere?  

#### Nice to haves

These options are categorised as _nice to have_ meaning that if someone can see an easy way to do it we can add them but they should not block the deployment of the gallery.

- [ ] this database could also include a) where the images are in the book, b) when they were created, c) any other metadata.
  * Probably best saved for a follow up issue unless someone can add an easy line or two of code to the pull request.
- [ ] Can the images be categorised and linked together? See discussion below at https://github.com/the-turing-way/the-turing-way/issues/2046#issuecomment-2457241692
- [ ] display metadata like: where in the book the illustration is featured (or not), which DOI or version it is associated it 
  * use case: DOIs for citing, (re)discovering earlier illustrations which may not have alt-text

### Who can help?

* Anyone who knows a good solution can suggest
* Anyone who has time and interest in addressing this, can take on deploying the solution
  * Thank you to @nehamoopen who is taking a first pass!

---

#### History

* A slack conversation between Me, @sgibson91 and @RaoOfPhysics took place (seeded by a discussion with @AidaMehonic and @KirstieJane)
* @EstherPlomp also expressed the need
* @nehamoopen worked on this before and during the November 2024 book dash

**Comment by malvikasharan** [comment]
- Suggested by @sgibson91: 
  - https://github.com/alexwlchan/docstore
  - https://github.com/briankosw/nbimageviewer
- @RaoOfPhysics suggested doing a feature request for Zenodo
- @EstherPlomp shared this repo: https://github.com/allisonhorst/stats-illustrations

**Comment by aleesteele** [comment]
Adding these here as well: 
- https://betterimagesofai.org/
- https://alan-turing-institute.github.io/turing-commons/resources/gallery/

**Comment by nehamoopen** [comment]
Here's a prototype of an illustration gallery built using R Shiny: https://nehamoopen.shinyapps.io/the-turing-way-illustrations/

![image](https://github.com/user-attachments/assets/cc99b471-a846-4232-ba35-42a55618fd3d)

This started off as a side-project to improve programming skills and I'm excited to develop it further in collaboration with others! I'm currently at the Book Dash (November 2024) and inviting feedback from Book Dash attendees and the broader community. At this stage, I want to create some kind of roadmap for the gallery that will provide (development) direction for me and/or potential collaborators.

Here's an overview of the ideas & feedback so far:

- Keyword search
- Alt-text 
- Download button 
- Scriberia attribution 
- License 
- Will it be embedded in the book or hosted elsewhere?  
- Does it need to be built in Python instead of R?
- Auto resize images to fit screen
- Click through buttons for images (back and forth, like slides)
- Inspiration: https://nsgrantham.shinyapps.io/tidytuesdayrocks/ 

More ideas & feedback are welcome! Eventually, it might be good to prioritize to-dos...

**Comment by SaranjeetKaur** [comment]
I am wondering if it would help to categorise the illustrations more broadly - so that if someone selects one category then they get to view all the illustrations from that category. What do you think?

**Comment by nehamoopen** [comment]
> I am wondering if it would help to categorise the illustrations more broadly - so that if someone selects one category then they get to view all the illustrations from that category. What do you think?

I think this could be linked to implementing a keyword search. The Shiny components _select_ and _selectize_ could come in handy for: https://shiny.posit.co/r/components/

I'm not yet sure how it would go with rendering and displaying the images though. We now have one image displayed at a time. We could take inspiration from this - https://nsgrantham.shinyapps.io/tidytuesdayrocks/ - where (a subselection of) single images can be clicked through. But if we want a gallery with multiple images at a time, we might need more Shiny expertise or we'll need to think about another solution.

**Comment by nehamoopen** [comment]
- [ ] show a random illustration
- [ ] display metadata like where in the book the illustration is featured (or not)
- [ ] categories, this may require the community to add tags (if not already available)

**Comment by nehamoopen** [comment]
- [ ] build a spreadsheet/database of images with metadata and build the gallery with that eventually

**Comment by nehamoopen** [comment]
Note: keep up with #3882 about adding alt-text

**Comment by nehamoopen** [comment]
## Roadmap

Based on the input received during the Book Dash (November 2024), here's a rough roadmap for developing the app further:

### v1.1 🚀

- [ ] welcome text
- [X] keyword search
- [ ] scriberia attribution
- [ ] download link
- [ ] license

### v1.2 🤩

- [ ] centralized document with figure names + alt-text
- [ ] implement alt-text

### v2.1 💡

- [ ] include images without text
- [ ] pagination of images
- [ ] button for random image

### v2.2 🔁

- [ ] fully automate deployment with GitHub Actions
- [ ] chapter in The Turing Way linking to the app?
- [ ] transfer repo to The Turing Way GitHub organization?

### v? 😱

- [ ] category/tag-based search + display of results

**Comment by RichardJActon** [comment]
This might be worth a look in the future for thinking about the image gallery feature: https://restlessdata.au/p/shiny-image-gallery-examples/

**Comment by nehamoopen** [comment]
> This might be worth a look in the future for thinking about the image gallery feature: https://restlessdata.au/p/shiny-image-gallery-examples/

Thanks @RichardJActon! That's one of the resources I found. I think there was a package being developed for the code used in this example, but I'm not sure how up to date it is. I'll be diving into it!

**Comment by nehamoopen** [comment]
I'm still Googling around and the pixture package might be in line with what we're looking for: https://www.royfrancis.com/pixture/

**Comment by JimMadge** [comment]
As we are looking to move to JupyterBook2/MystMD,

I've been looking a bit at MystMD internals and creating roles/directives. It would be quite easy to iterate over all figures in the book. So, I think creating a role to index all of the figures in the book should be quite straightforward and would likely be useful to other MystMD users (if it doesn't already exist).

**Comment by agoose77** [comment]
@JimMadge creating a custom directive to generate a gallery is a common pattern that we are seeing emerge: 

- source: https://github.com/projectpythia-mystmd/cookbook-gallery/blob/refs/heads/main/root.md
- plugin: https://github.com/projectpythia-mystmd/cookbook-gallery/blob/main/pythia-gallery.py

This plugin is a _Python_ plugin, a feature that the `mystmd` team prototyped during SciPy 2024. Most plugins should be JavaScript plugins, because they're easier to share and don't involve spinning up another interpreter. But, the core logic can be re-written to target JS instead of Python.

A more feature-rich gallery, such as one that can implement filtering etc. is something we need to think about as an ecosystem

**Comment by sayalaruano** [comment]
Hi @JimMadge, do you know if someone is working on creating the role to index the figures and try to build a gallery similar to the example provided by @agoose77? I'm looking for ideas to participate in May's BookDash, and this would be an interesting one.

**Comment by agoose77** [comment]
@sayalaruano on the MyST side, we're thinking about galleries generally, and it is a feature that 2i2c will likely be working on in the near future: https://github.com/2i2c-org/infrastructure/issues/5341

The approach that we're thinking of at the moment is one in which the view aspect of the gallery is handled naturally as MyST AST (i.e., it should be out-of-the-box MyST Markdown syntax), and then the filtering and sorting part is a theme (TypeScript) feature. As such, I think it would be reasonable to work on something like this -- anything that the JB team or 2i2c work on will likely work with what you'd be doing here.

**Comment by JimMadge** [comment]
Hi @sayalaruano, I don't think anyone is working on this from the Turing Way side. It would be great for someone to have a look into what is possible.

@agoose77 is it worth @sayalaruano linking up with the MyST team so we aren't duplicating effort? I think from a _project_ perspective I'd be happy to wait for a gallery feature in MyST, so wouldn't push to make our own plugin in the interim if we are likely to abandon it in the future.

**Comment by agoose77** [comment]
@JimMadge at the moment, every example of a gallery that I'm aware of sounds fairly bespoke. Unless that changes, the best path forward near-term would be to treat the MyST gallery "feature" as an enhancement that adds a dynamic view for filtering content (things like tags, search, etc). The idea would then be for e.g. TTW to generate the AST for e.g. a grid of images, that has some annotation with e.g. tags that the theme enables filtering against.

So, TL;DR --- go for it, I think we'll likely be moving in that direction.

**Comment by sayalaruano** [comment]
Hey!  During this Book Dash May 2025, I created a [MyST Python executable plugin](https://github.com/sayalaruano/myst-image-gallery) to create an image gallery. Here is an [example](https://sayalaruano.github.io/myst-image-gallery/) of how it looks. Based on this, I opened a PR to include a gallery as a subchapter of the TW book: https://github.com/the-turing-way/the-turing-way/pull/4224

For some reason, in the TW book, the gallery is not showing up, so I need to still debug it to make it work. Feel free to take a look, and your suggestions and feedback are more than welcome.

**Comment by agoose77** [comment]
@sayalaruano this is very cool! I would love to hear about your experience of authoring this plugin, is this something you've already written about anywhere? If not, would you be interested in joining our next Jupyter Book meeting to share your experience?

**Comment by sayalaruano** [comment]
Hey @agoose77! I have not written about the process to create the plugin yet. And yes, it would be cool to join the Jupyter Book meeting. You can send me an invitation to s_ayalaruano@proton.me

**Comment by nehamoopen** [comment]
Hello! I'm closing this issue because the infrastructure discussed in #4224 (a plugin for MyST) can be directly integrated with the book and maintained by the community, whereas my side-project was built with R and I haven't worked on it much since I proposed it. So please head to #4224 and help out there! ✨
